### PR TITLE
chore(TDKN-217): Add token based authentication support

### PR DESCRIPTION
**Pull request by fhuaulme**
**What is the problem this Pull Request is trying to solve?**
See https://jira.talendforge.org/browse/TDKN-217 and relates to https://github.com/Talend/policies/pull/35
 
**What is the chosen solution to this problem?**
See https://jira.talendforge.org/browse/TDKN-217
 
**Link to the JIRA issue**
https://jira.talendforge.org/browse/TDKN-217
 
**Please check if the Pull Request fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/Talend/daikon/blob/master/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features, coverage should be over 75% in the new code)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in JIRA), if any, are all linked or available in the Pull Request

### CODE REVIEW

File: daikon-spring/daikon-spring-security/README.adoc [hunk 1]
```diff
@@ -1,20 +1,34 @@
...
+- link:https://spring.io/guides/gs/actuator-service/[Actuator]
+- link:https://[Version information]
+
+When this module is added to application classpath, all the endpoints above are secured with a token in configuration.
+
+```
+talend.security.token.value=SeCr3tT0keNVa1ue
+```
+
+All HTTP requests to Actuator or Version must then include the HTTP `X_TALEND_ADMIN_TOKEN` (HTTP header can be configured using `talend.security.token.http.header` in configuration).
```
Comment (rdalverny) [inline daikon-spring/daikon-spring-security/README.adoc]: I would reuse HTTP header `Authorization`, with a `Authorization: Talend <token-value>` format; and not let it configurable here.

File: daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    public TokenAuthenticationFilterTest() throws IOException, ServletException {
+        filterChain = mock(FilterChain.class);
+        doAnswer(invocationOnMock -> {
+            capturedContext.set(SecurityContextHolder.getContext());
+            return null;
+        }).when(filterChain).doFilter(any(), any());
+        response = mock(HttpServletResponse.class);
+        request = mock(HttpServletRequest.class);
+
+        filter = new TokenAuthenticationFilter(DigestUtils.sha256Hex("myToken"), "X_TALEND_TOKEN", "AUTH1, AUTH2");
```
Comment (rdalverny) [inline daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationFilterTest.java]: Not consistent with above doc (missing an `_ADMIN`; other instances of the same below).

File: daikon-spring/daikon-spring-security/README.adoc (hunk 1 above)
Comment (fhuaulme) [reply daikon-spring/daikon-spring-security/README.adoc]: I didn't have the complete spec at the PR was pushed. Change to come.

File: daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationFilterTest.java (hunk 2 above)
Comment (fhuaulme) [reply daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationFilterTest.java]: It isn't. The unit test can use the configuration it wants and don't rely on Spring configuration keys.

However, unit test is going to change, making this obsolete.

File: daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationFilterTest.java [hunk 3]
```diff
@@ -66,7 +66,7 @@ public void shouldRejectRequestWithNoHeader() throws IOException, ServletExcepti
     @Test
     public void shouldAllowRequestWithValidToken() throws IOException, ServletException {
         // given
-        when(request.getHeader(eq("X_TALEND_TOKEN"))).thenReturn("myToken");
+        when(request.getHeader(eq("Authentication"))).thenReturn("Talend myToken");
```
Comment (rdalverny) [inline daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationFilterTest.java]:
```suggestion
        when(request.getHeader(eq("Authorization"))).thenReturn("Talend myToken");
```
And below as well.

File: daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationFilterTest.java [hunk 4]
```diff
@@ -77,14 +77,27 @@ public void shouldAllowRequestWithValidToken() throws IOException, ServletExcept
...
+            return "ROLE_ACTUATOR".equals(authority) || "ROLE_VERSION".equals(authority);
         }));
     }
 
     @Test
-    public void shouldRejectRequestWithInvalidToken() throws IOException, ServletException {
+    public void shouldRejectRequestWithInvalidTokenFormat() throws IOException, ServletException {
         // given
-        when(request.getHeader(eq("X_TALEND_TOKEN"))).thenReturn("invalidToken");
+        when(request.getHeader(eq("Authentication"))).thenReturn("xxx");
```
Comment (rdalverny) [inline daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationFilterTest.java]: Not sure, maybe Springboot makes all headers uppercase?

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/158/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/158/)

File: daikon-spring/daikon-spring-security/pom.xml [hunk 5]
```diff
@@ -1,69 +1,72 @@
...
+        <dependencies>
+            <dependency>
+                <groupId>org.springframework.boot</groupId>
+                <artifactId>spring-boot-dependencies</artifactId>
+                <version>${spring-boot.version}</version>
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+            <dependency>
+                <groupId>org.springframework.cloud</groupId>
```
Comment (sgandon) [inline daikon-spring/daikon-spring-security/pom.xml]: what does this dependency bring ?

File: daikon-spring/daikon-spring-security/pom.xml [hunk 6]
```diff
@@ -1,69 +1,72 @@
...
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-aop</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.springframework</groupId>
-            <artifactId>spring-test</artifactId>
-            <version>${spring.version}</version>
-            <scope>test</scope>
+            <groupId>io.micrometer</groupId>
+            <artifactId>micrometer-registry-prometheus</artifactId>
```
Comment (sgandon) [inline daikon-spring/daikon-spring-security/pom.xml]: why would you need prometheuse here in the lib ?

Comment (sgandon) [reply daikon-spring/daikon-spring-security/pom.xml]: As far as I see it, it is used in the test, so I guess this dependency should be scoped with "test"

File: daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationTest.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void shouldNotAllowAuthenticatedChange() {
+        authentication.setAuthenticated(false);
+    }
+
+    @Test
+    public void shouldAllowAuthenticatedEnable() {
+        authentication.setAuthenticated(true);
```
Comment (sgandon) [inline daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationTest.java]: what does this test do ?

Comment (sgandon) [reply daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationTest.java]: I guess it should not throw any exception, may you could write some comment here to explain it.

File: daikon-spring/daikon-spring-security/pom.xml (hunk 5 above)
Comment (fhuaulme) [reply daikon-spring/daikon-spring-security/pom.xml]: Nothing, it's _slightly_ overkill :). Removing it

File: daikon-spring/daikon-spring-security/pom.xml (hunk 6 above)
Comment (fhuaulme) [reply daikon-spring/daikon-spring-security/pom.xml]: Indeed, only needed for showcase. Removing it.

File: daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationTest.java (hunk 7 above)
Comment (fhuaulme) [reply daikon-spring/daikon-spring-security/src/test/java/org/talend/daikon/security/token/TokenAuthenticationTest.java]: Adding comment to explain this test.

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/159/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/159/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/160/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/160/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/161/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/161/)

File: daikon-spring/daikon-spring-security/pom.xml [hunk 8]
```diff
@@ -1,70 +1,62 @@
...
-            <scope>provided</scope>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-aop</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.springframework</groupId>
-            <artifactId>spring-context</artifactId>
-            <version>${spring.version}</version>
+            <groupId>org.talend.daikon</groupId>
+            <artifactId>daikon</artifactId>
```
Comment (sgandon) [inline daikon-spring/daikon-spring-security/pom.xml]:
```suggestion
            <artifactId>daikon-exception</artifactId>
```
I have tested it locally and only this module is required.

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/162/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/162/)

File: daikon-spring/daikon-spring-security/pom.xml (hunk 8 above)
Comment (fhuaulme) [reply daikon-spring/daikon-spring-security/pom.xml]: 👍

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/163/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/163/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/164/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/164/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/166/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/166/)

**Pull request merged**